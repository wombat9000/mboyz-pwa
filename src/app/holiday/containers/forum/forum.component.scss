@import '../../../../../node_modules/@angular/material/theming';
@import '../../../../mboyz-theme';

:host {
  display: block;
  padding: 8px;

  .message-input {
    width: 100%;
    margin-top: 5px;
  }

  .posts-container {
    .post {
      width: 100%;
      background-color: mat-color($mboyz-accent, 100);
      margin-bottom: 10px;
    }
  }
}
