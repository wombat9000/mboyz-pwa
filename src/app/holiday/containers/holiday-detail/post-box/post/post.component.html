<div class="post-body" *ngIf="user$ | async as user">
  <div class="body">
    <span class="author">{{user.displayName}}</span>
    <span class="message">{{post.text}}</span>
  </div>
  <div class="info">
    <span class="created-text">{{formatDate(post.created)}} um {{formatTime(post.created)}}</span>
  </div>
</div>
<div class="comments">
  <app-comment-box [post]="post"></app-comment-box>
</div>
