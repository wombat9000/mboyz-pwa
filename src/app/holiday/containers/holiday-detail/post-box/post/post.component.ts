import {Component, Input, OnInit} from '@angular/core';
import {Observable} from 'rxjs/Observable';
import * as moment from 'moment';
import {Post} from '../../../../services/post-firestore.service';
import {MtravelUser} from '../../../../../auth/services/auth.service';
import {UserFirestore} from '../../../../../auth/services/user-firestore.service';


@Component({
  selector: 'app-post',
  templateUrl: './post.component.html',
  styleUrls: ['./post.component.scss']
})
export class PostComponent implements OnInit {

  @Input()
  post: Post;

  user$: Observable<MtravelUser>;

  constructor(private userFS: UserFirestore) {
  }

  ngOnInit() {
    this.user$ = this.userFS.observeById(this.post.authorId);
  }

  formatDate(isoString: string): string {
    return moment(isoString).format('Do MMMM');
  }

  formatTime(isoString: string): string {
    return moment(isoString).format('LT');
  }
}
