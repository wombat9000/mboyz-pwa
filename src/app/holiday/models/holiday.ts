import {now} from 'moment';
import {DbRecord} from './DbRecord';

export interface HolidayDTO extends DbRecord {
  readonly name: string;
  readonly created: string;
}

export function newTestHoliday(id: string): HolidayDTO {
  return {
    id: id,
    name: 'someName',
    created: now().toString()
  };
}
