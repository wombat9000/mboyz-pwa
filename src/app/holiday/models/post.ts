import {DbRecord} from './DbRecord';

export interface PostDTO extends DbRecord {
  readonly text: string;
  readonly holidayId: string;
  readonly authorId: string;
  readonly created: string;
}
