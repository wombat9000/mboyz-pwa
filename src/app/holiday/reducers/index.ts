import {createFeatureSelector, createSelector} from '@ngrx/store';
import * as fromHolidays from './holiday.reducer';
import * as fromRoot from '../../reducers';


export interface HolidaysState {
  holidays: fromHolidays.State;
}

export interface State extends fromRoot.State {
  holidays: HolidaysState;
}

export const reducers = {
  holidays: fromHolidays.reducer
};

export const getHolidaysState = createFeatureSelector<HolidaysState>('holidays');

export const getHolidayEntitiesState = createSelector(
  getHolidaysState,
  state => state.holidays
);

export const getSelectedHolidayId = createSelector(
  getHolidayEntitiesState,
  fromHolidays.getSelectedId
);

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = fromHolidays.adapter.getSelectors(getHolidayEntitiesState);

export const getSelectedHoliday = createSelector(
  selectEntities,
  getSelectedHolidayId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);
