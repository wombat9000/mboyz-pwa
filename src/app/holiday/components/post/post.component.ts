import {Component, Input} from '@angular/core';
import * as moment from 'moment';
import {MtravelUser} from '../../../auth/services/auth.service';
import {PostDTO} from '../../models/post';
import {CommentDTO} from '../../models/comment';
import * as uuid from 'uuid';
import {Create} from '../../actions/comment.actions';
import {Store} from '@ngrx/store';
import * as fromRoot from '../../../reducers';

@Component({
  selector: 'app-post',
  template: `
    <div class="post-body">
      <div class="body">
        <span class="author">{{author.displayName}} </span>
        <span class="message">{{post.text}}</span>
      </div>
      <div class="info">
        <span class="created-text">{{formatDate(post.created)}} um {{formatTime(post.created)}}</span>
      </div>
    </div>
    <div class="comments">
      <app-comment-box [post]="post"
                       [comments]="comments"
                       [activeUser]="activeUser"
                       (submitComment)="submitComment($event)">
      </app-comment-box>
    </div>
  `,
  styleUrls: ['./post.component.scss']
})
export class PostComponent {

  @Input()
  post: PostDTO;
  @Input()
  comments: CommentDTO[];
  @Input()
  author: MtravelUser;
  @Input()
  activeUser: MtravelUser;

  constructor(private store: Store<fromRoot.State>) {
  }

  submitComment(text: string) {
    const comment: CommentDTO = {
      id: uuid(),
      text: text,
      postId: this.post.id,
      holidayId: this.post.holidayId,
      authorId: this.activeUser.id,
      created: moment().toISOString()
    };

    this.store.dispatch(new Create({record: comment}));
  }

  formatDate(isoString: string): string {
    return moment(isoString).format('Do MMMM');
  }

  formatTime(isoString: string): string {
    return moment(isoString).format('LT');
  }
}
