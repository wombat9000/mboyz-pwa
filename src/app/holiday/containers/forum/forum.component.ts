import {Component, Input} from '@angular/core';
import * as moment from 'moment';
import {animate, keyframes, style, transition, trigger} from '@angular/animations';
import {HolidayDTO} from '../../models/holiday';
import {MtravelUser} from '../../../auth/services/auth.service';
import {PostDTO} from '../../models/post';
import {Store} from '@ngrx/store';
import * as fromRoot from '../../../reducers/index';
import {Create} from '../../actions/post.actions';
import * as uuid from 'uuid';


@Component({
  selector: 'app-forum',
  template: `
    <mat-form-field class="message-input">
  <textarea matInput
            matTextareaAutosize
            data-qa="post-input"
            (keyup.enter)="submitPost()"
            [(ngModel)]="postInput"
            placeholder="Neuer Beitrag ..."></textarea>
    </mat-form-field>
    <div class="posts-container">
      <div class="post" *ngFor="let post of posts" [@fadeIn]>
        <app-post [post]="post"
                  [activeUser]="activeUser">
        </app-post>
      </div>
    </div>
  `,
  styleUrls: ['./forum.component.scss'],
  animations: [
    trigger('fadeIn', [
      transition('void => *', [
        animate('0.3s ease-in', keyframes([
          style({opacity: 0}),
          style({opacity: 1})
        ]))
      ])
    ])
  ]
})
export class ForumComponent {

  @Input()
  holiday: HolidayDTO;
  @Input()
  activeUser: MtravelUser;
  @Input()
  posts: PostDTO[];

  postInput = '';

  constructor(private store: Store<fromRoot.State>) {
  }

  submitPost() {
    const post: PostDTO = {
      id: uuid(),
      text: this.postInput,
      holidayId: this.holiday.id,
      authorId: this.activeUser.id,
      created: moment().toISOString()
    };

    this.store.dispatch(new Create({record: post}));
    this.postInput = '';
  }
}
