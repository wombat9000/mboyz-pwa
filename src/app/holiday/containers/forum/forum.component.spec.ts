import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {CUSTOM_ELEMENTS_SCHEMA, DebugElement} from '@angular/core';
import {By} from '@angular/platform-browser';
import {FormsModule} from '@angular/forms';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {AuthService, MtravelUser} from '../../../auth/services/auth.service';
import {authServiceMocker} from '../../../test-support/stubs';
import {HolidayDTO} from '../../models/holiday';
import {PostDTO} from '../../models/post';
import {combineReducers, Store, StoreModule} from '@ngrx/store';
import * as fromHoliday from '../../reducers/index';
import {HolidaysState} from '../../reducers';
import {of} from 'rxjs';
import {ForumComponent} from './forum.component';
import moment = require('moment');

class PostBoxPO {
  constructor(private fixture: ComponentFixture<ForumComponent>,
              holiday: HolidayDTO,
              activeUser: MtravelUser,
              posts: PostDTO[]) {
    const component = fixture.componentInstance;
    component.holiday = holiday;
    component.activeUser = activeUser;
    component.posts = posts;
  }

  postInput() {
    return this.fixture.debugElement.query(By.css('textarea'));
  }

  createPost(text: string) {
    const input = this.postInput();
    input.nativeElement.value = text;
    input.nativeElement.dispatchEvent(new Event('input'));
    input.nativeElement.dispatchEvent(new KeyboardEvent('keyup', {
      'key': 'Enter'
    }));
    this.fixture.detectChanges();
  }
}

describe('ForumComponent', () => {

  const someAuthor: MtravelUser = {
    displayName: 'Pinky Floyd',
    id: 'someUid',
    email: 'someMail',
    photoURL: null
  };

  let fixture: ComponentFixture<ForumComponent>;
  let debugElement: DebugElement;
  let authServiceMock: jasmine.SpyObj<AuthService>;
  let postBoxPO: PostBoxPO;
  let store: Store<HolidaysState>;

  const inputHoliday = {
    id: 'someId',
    name: 'someName',
    created: ''
  };

  const somePost: PostDTO = {
    id: 'someId',
    text: 'first message',
    holidayId: 'holidayId',
    authorId: 'someAuthor',
    created: moment('2016-01-01').toISOString()
  };

  const moreRecentPost: PostDTO = {
    id: 'anotherId',
    text: 'second message',
    holidayId: 'holidayId',
    authorId: 'someAuthor',
    created: moment('2016-01-02').toISOString()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        FormsModule, NoopAnimationsModule,
        StoreModule.forRoot({
          holidayPlanner: combineReducers(fromHoliday.reducers),
        })
      ],
      providers: [
        {provide: AuthService, useFactory: authServiceMocker}
      ],
      declarations: [ForumComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    authServiceMock = TestBed.get(AuthService);
    store = TestBed.get(Store);

    authServiceMock.activeUser.mockReturnValue(of(someAuthor));

    fixture = TestBed.createComponent(ForumComponent);
    postBoxPO = new PostBoxPO(fixture, inputHoliday, someAuthor, [somePost, moreRecentPost]);

    debugElement = fixture.debugElement;
    fixture.detectChanges();
  });

  describe('displays posts', () => {
    let postedMessages: string[];

    beforeEach(async () => {
      await fixture.whenStable();
      fixture.detectChanges();

      postedMessages = debugElement.queryAll(By.css('app-post'))
        .map(it => it.properties.post.text);
    });

    it('posts should appear in the list', () => {
      expect(postedMessages).toContain(somePost.text);
      expect(postedMessages).toContain(moreRecentPost.text);
    });
  });

  describe('creating a new post', () => {
    let post: PostDTO;
    const someMessage = 'someMessage';

    beforeEach(async () => {
      const dispatchSpy = spyOn(store, 'dispatch');

      postBoxPO.createPost(someMessage);
      await fixture.whenStable();
      fixture.detectChanges();

      post = dispatchSpy.calls.argsFor(0)[0].payload.record;
    });

    it('should clear the input field after the message is sent', () => {
      expect(fixture).toMatchSnapshot();
    });

    it('should persist the new message', () => {
      expect(post.text).toBe(someMessage);
    });

    it('should use currently logged in users name as someAuthor', () => {
      expect(post.authorId).toBe(someAuthor.id);
    });
  });
});
