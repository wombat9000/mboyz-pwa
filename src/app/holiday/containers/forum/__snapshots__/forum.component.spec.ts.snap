// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForumComponent creating a new post should clear the input field after the message is sent 1`] = `
<app-forum
  activeUser={[Function Object]}
  holiday={[Function Object]}
  postInput=""
  posts={[Function Array]}
  store={[Function Store]}
>
  <mat-form-field
    class="message-input"
  >
    <textarea
      class="ng-tns-c1-0 ng-untouched ng-valid ng-dirty"
      data-qa="post-input"
      matinput=""
      mattextareaautosize=""
      ng-reflect-model=""
      placeholder="Neuer Beitrag ..."
    />
  </mat-form-field><div
    class="posts-container"
  >
    
    <div
      class="post ng-tns-c1-0 ng-trigger ng-trigger-fadeIn ng-star-inserted"
    >
      <app-post
        class="ng-tns-c1-0"
      />
    </div>
    <div
      class="post ng-tns-c1-0 ng-trigger ng-trigger-fadeIn ng-star-inserted"
    >
      <app-post
        class="ng-tns-c1-0"
      />
    </div>
  </div>
</app-forum>
`;
